### PR TITLE
BehaviorSpace disable plots when running plots headlessly

### PR DESCRIPTION
**Pull request by ryanp8**
**Motivation**
Fixes issue #2120. 

**Description**
In the Run Options, there is a checkbox to enable or disable updating plots. When this box is unchecked, plots are disabled for GUI and headless workspaces. The corresponding checkbox in the Running Experiment dialog is then disabled. If the Run Options box is checked, then GUI and headless plots are updated, and the user can toggle the corresponding checkbox in "Running Experiment"

When running the Red Queen Beta2 experiment for 500,000 ticks, the performance is similar to the workaround described in the original issue.
image

**Other Changes**
- A new option, `--update-plots`, has been added when running experiments from the console.
- The user is now given an explanatory warning when `export-interface`, `export-plot`, `export-all-plots`, or `export-world` are used when plots are disabled. 
- The documentation has also been updated to reflect the changes.

### CODE REVIEW

**Comment by brandesNW** [comment]
Currently error messages such as "You are trying to export a plot, but plot updates are disabled." are issued for each run.  There should only be one. Also I think it should be a non-blocking message for which you have a "dismiss" button (example - error dialog for a runtime error such as trying to export a plot that doesn't exist)

**Comment by brandesNW** [comment]
ExportPlotWarningAction.Throw doesn't seem to be used and can be removed

**Review by brandesNW — CHANGES_REQUESTED** [review]

File: netlogo-core/src/main/lab/Worker.scala [hunk 1]
```diff
@@ -192,6 +192,39 @@ class Worker(val protocol: LabProtocol)
...
+          if (!exportCommandFound) exportCommandFound = checkForPlotExportCommand(goProcedure.code)
+          if (!exportCommandFound) exportCommandFound = checkForPlotExportCommand(setupProcedure.code)
+
+          if (exportCommandFound) {
+            import ExportPlotWarningAction._
+            ws.setTriedToExportPlot(true)
+            ws.exportPlotWarningAction match {
+              case Output => {
+                ws.setExportPlotWarningAction(ExportPlotWarningAction.Ignore)
+                println("Enable plot updating in Run Options to use export-plot, export-all-plots, export-world, or export-interface.")
```
Comment (LaCuneta) [inline netlogo-core/src/main/lab/Worker.scala:221]: Line is over 120 characters long, maybe move it to an error I18N string?

File: netlogo-headless/src/main/workspace/AbstractWorkspace.scala [hunk 2]
```diff
@@ -221,6 +238,21 @@ object AbstractWorkspaceTraits {
     def exportDrawingToCSV(writer:PrintWriter)
     def exportOutputAreaToCSV(writer:PrintWriter)
 
+    def checkPlotUpdates() {
+      import ExportPlotWarningAction._
+      exportPlotWarningAction match {
+        case Warn => {
+          setExportPlotWarningAction(ExportPlotWarningAction.Ignore)
+          throw new Exception("Enable plot updating to use export-plot, export-all-plots, export-world, or export-interface.")
```
Comment (LaCuneta) [inline netlogo-headless/src/main/workspace/AbstractWorkspace.scala:246]: Probably move to I18N string, merged with the other similar usages.

File: netlogo-gui/src/main/lab/gui/ProgressDialog.scala [hunk 3]
```diff
@@ -213,7 +219,26 @@ private [gui] class ProgressDialog(dialog: java.awt.Dialog, supervisor: Supervis
...
+                OptionDialog.showMessage(
+                  workspace.getFrame, "Updating Plots Warning",
+                  "Warning: enable plot updating in Run Options if you want to collect plot data using export-plot, export-all-plots, export-world, or export-interface.",
+                  Array(I18N.gui.get("common.buttons.continue"))
+                )
+              }
+            })
+          }
+          case _ =>
+        }
```
Comment (LaCuneta) [inline netlogo-gui/src/main/lab/gui/ProgressDialog.scala:239]: Can simplify to the below without a single-case `match`, and also merge the error message with the errors as an I18N string.

```scala
      if (workspace.triedToExportPlot && workspace.exportPlotWarningAction == ExportPlotWarningAction.Warn) {
        workspace.setExportPlotWarningAction(ExportPlotWarningAction.Ignore)
        org.nlogo.awt.EventQueue.invokeLater(new Runnable() {
          def run() {
            OptionDialog.showMessage(
              workspace.getFrame, "Updating Plots Warning",
              "Warning: enable plot updating in Run Options if you want to collect plot data using export-plot, export-all-plots, export-world, or export-interface.",
              Array(I18N.gui.get("common.buttons.continue"))
            )
          }
        })
      }
```

File: netlogo-gui/src/main/lab/gui/RunOptionsDialog.scala [hunk 4]
```diff
@@ -87,7 +87,7 @@ class RunOptionsDialog(parent: java.awt.Dialog,
         Property("spreadsheet", Property.FilePath(spreadsheetFile), I18N.gui("spreadsheet")),
         Property("table", Property.FilePath(tableFile), I18N.gui("table")),
         Property("updateView", Property.Boolean, I18N.gui("updateview")),
-        Property("updatePlotsAndMonitors", Property.Boolean, I18N.gui("updateplotsandmonitors")),
+        Property("updatePlotsAndMonitors", Property.Boolean, I18N.gui("updateplotsandmonitors"), "<html>" + I18N.gui("updateplotsandmonitors.info") + "</html>"),
```
Comment (LaCuneta) [inline netlogo-gui/src/main/lab/gui/RunOptionsDialog.scala:90]: Nit: Prefer lines stay under 120 characters.

File: netlogo-headless/src/main/workspace/AbstractWorkspace.scala [hunk 5]
```diff
@@ -221,6 +238,21 @@ object AbstractWorkspaceTraits {
...
+        }
+        case Output => {
+          setExportPlotWarningAction(ExportPlotWarningAction.Ignore)
+          println("Enable plot updating to use export-plot, export-all-plots, export-world, or export-interface.")
```
Comment (LaCuneta) [inline netlogo-headless/src/main/workspace/AbstractWorkspace.scala:250]: Same as above.

**Review by LaCuneta — CHANGES_REQUESTED** [review]
Nice, thank you for this.  It's going to help prevent a lot of wasted CPU and IRL time.  I left some small items in the comments.

One item I encountered was that the **Run Options** dialog looks bizarre on my system, it's squished and some of the help lines seem mis-aligned (screenshot below).  I'm on Ubuntu 22.04 using the older Xorg display server, if it matters.

Run options window scrunched

File: netlogo-gui/src/main/headless/Main.scala [hunk 6]
```diff
@@ -189,6 +197,8 @@ See the Advanced Usage section of the BehaviorSpace documentation in the NetLogo
       } else if (arg == "--suppress-errors") {
         suppressErrors = true
 
+      } else if (arg == "--update-plots") {
```
[the remark this answers is not in the document]
Comment (LaCuneta) [reply netlogo-gui/src/main/headless/Main.scala:200]: Also forgot to mention: we should add this to the big help blurb string at the top of this file.

[the remark this answers is not in the document]
Comment (ryanp8) [reply netlogo-gui/src/main/headless/Main.scala:201]: Done

File: netlogo-core/src/main/lab/Worker.scala (hunk 1 above)
Comment (ryanp8) [reply netlogo-core/src/main/lab/Worker.scala]: Done

File: netlogo-headless/src/main/workspace/AbstractWorkspace.scala (hunk 2 above)
Comment (ryanp8) [reply netlogo-headless/src/main/workspace/AbstractWorkspace.scala]: Done

File: netlogo-gui/src/main/lab/gui/ProgressDialog.scala (hunk 3 above)
Comment (ryanp8) [reply netlogo-gui/src/main/lab/gui/ProgressDialog.scala]: Done

File: netlogo-headless/src/main/workspace/AbstractWorkspace.scala (hunk 5 above)
Comment (ryanp8) [reply netlogo-headless/src/main/workspace/AbstractWorkspace.scala]: Done

**Comment by LaCuneta** [comment]
@ryanp8 Thanks for the updates!  Just FYI you can use the **Resolve Conversation** button on a comment to mark it, well, resolved.  It makes it easier to track what's done and left to do.

The only thing still not quite working right is the display of the run dialog.  You can see in the screenshot I left how it's squished together and the help text line breaks are off.  Once that is resolved I think this can be merged.

File: netlogo-gui/src/main/properties/EditPanel.scala [hunk 7]
```diff
@@ -62,6 +62,7 @@ class EditPanel(val target: Editable, val compiler: CompilerServices, colorizer:
       c.anchor = GridBagConstraints.WEST
       c.insets = new Insets(3, 3, 3, 3)
       c.gridwidth = property.gridWidth
+      if (editor.isInstanceOf[BooleanEditor]) c.fill = GridBagConstraints.HORIZONTAL
```
Comment (LaCuneta) [inline netlogo-gui/src/main/properties/EditPanel.scala:65]: This did fix the squishing issue, but ideally the editor panel shouldn't have to know anything about how the specific property editors are setup.  Any way to get the `HORIZONTAL` constraint setting moved into the `BooleanEditor` itself?

**Review by LaCuneta — APPROVED** [review]
Nice, works for me!

**Pull request merged**